src/build.rs

```diff
@@ -422,63 +421,153 @@ impl BuildQueue {
+                        let (package, all) = match config.analyze_package {
+                            Some(ref pkg_name) => (vec![pkg_name.clone()], false),
+                            None => (vec![], true),
+                        };
+
+                        CargoOptions {
+                            flag_package: package,
+                            flag_all: all,
+                            flag_target: config.target.clone(),
+                            .. CargoOptions::default()
+                        }
+                    } else {
+                        // In non-workspace mode we currently support only one crate target,
+                        // and if lib is set, then we ignore bin target config
+                        let (lib, bin) = match config.build_lib {
+                            true => (true, vec![]),
+                            false => {
+                                let bin = match config.build_bin {
+                                    Some(ref bin) => vec![bin.clone()],
+                                    None => vec![],
+                                };
+                                (false, bin)
+                            },
+                        };
+
+                        CargoOptions {
+                            flag_lib: lib,
+                            flag_bin: bin,
+                            flag_target: config.target.clone(),
+                            .. CargoOptions::default()
+                        }
+                    }
+                }
+            }
+
 
             let mut shell = Shell::from_write(Box::new(BufWriter(out.clone())));
             shell.set_verbosity(Verbosity::Quiet);
             let config = make_cargo_config(&build_dir, shell);
-            let mut manifest_path = build_dir.clone();
-            manifest_path.push("Cargo.toml");
+
+            let manifest_path = build_dir.join("Cargo.toml");
             trace!("manifest_path: {:?}", manifest_path);
+            //let root = find_root_manifest_for_wd(options.flag_manifest_path, config.cwd())?;
             // TODO: Add support for virtual manifests and multiple packages 
             let ws = Workspace::new(&manifest_path, &config).expect("could not create cargo workspace");
-            let current_package = ws.current().unwrap(); 
-            let targets = current_package.targets();
-            let bins;
-            let target_string;
 
-            let opts = {
-                let rls_config = rls_config.lock().unwrap();
+            // Copy over relevant options without locking whole config, as compilation procedure
+            // will need those and we don't want to lock whole config file during that period
+            // TODO: Keep this structure cached and regenerate it on every relevant configuration change
+            let rls_config = rls_config.lock().unwrap().clone();
+
+            let opts = CargoOptions::new(&rls_config);
```

Comment:
I don't really get why you use the CargoOptions struct? It seems to just be adding boilerplate